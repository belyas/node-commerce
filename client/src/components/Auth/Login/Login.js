import React, { useState } from 'react';

import { concatClasses } from '../../../utils/helpers';
import useFormValidation from '../../../hooks/useFormValidation';
import { validateLoginAuth } from '../../../utils/validateAuth';
import loginAuth from '../../../apis/authLogin';
import styles from './Login.module.css';

const INITIAL_STATE = {
    email: '',
    password: '',
};

const Login = () => {
    let [isLogin, setLogin] = useState(true);
    let hideForm = { display: 'none' };
    const toggleLoginHandler = () => {
        setLogin(!isLogin);
    };
    const {
        values,
        onChangeHandler,
        onSubmitHanlder,
        onBlurHandler,
        errors,
        isSubmitting,
    } = useFormValidation(INITIAL_STATE, validateLoginAuth, loginAuth);

    return (
        <div className={concatClasses(styles.wrapper, styles.fadeInDown)}>
            <div className={concatClasses(styles.autoMg, styles.formContent)}>
                <form
                    onSubmit={onSubmitHanlder}
                    style={isLogin ? {} : hideForm}>
                    <h2
                        className={concatClasses(
                            styles.fadeIn,
                            styles.first,
                            styles.formTitle
                        )}>
                        Login
                    </h2>
                    <input
                        type="text"
                        className={concatClasses(styles.fadeIn, styles.second)}
                        name="email"
                        placeholder="Email address"
                        value={values.email}
                        onChange={onChangeHandler}
                        onBlur={onBlurHandler}
                        required
                    />
                    {errors.email && (
                        <p className="alert alert-danger">{errors.email}</p>
                    )}
                    <input
                        type="password"
                        className={concatClasses(styles.fadeIn, styles.third)}
                        name="password"
                        placeholder="Password"
                        value={values.password}
                        onChange={onChangeHandler}
                        onBlur={onBlurHandler}
                        required
                    />
                    {errors.password && (
                        <p className="alert alert-danger">{errors.password}</p>
                    )}
                    <input
                        type="submit"
                        className={concatClasses(styles.fadeIn, styles.fourth)}
                        value="Log In"
                        disabled={isSubmitting}
                    />
                </form>

                <form action="#" method="POST" style={isLogin ? hideForm : {}}>
                    <h2
                        className={concatClasses(
                            styles.fadeIn,
                            styles.first,
                            styles.formTitle
                        )}>
                        Signup
                    </h2>
                    <input
                        type="text"
                        className={concatClasses(styles.fadeIn, styles.second)}
                        name="firstname"
                        placeholder="First name"
                        required
                    />
                    <input
                        type="text"
                        className={concatClasses(styles.fadeIn, styles.second)}
                        name="lastname"
                        placeholder="Last name"
                        required
                    />
                    <input
                        type="text"
                        className={concatClasses(styles.fadeIn, styles.third)}
                        name="email"
                        placeholder="Email address"
                        required
                    />
                    <input
                        type="password"
                        className={concatClasses(styles.fadeIn, styles.third)}
                        name="password"
                        placeholder="Password"
                        required
                    />
                    <input
                        type="button"
                        className={concatClasses(styles.fadeIn, styles.fourth)}
                        value="Register"
                    />
                </form>

                <div className={concatClasses(styles.formFooter)}>
                    <a
                        href="#/"
                        className={concatClasses(
                            styles.underlineHover,
                            styles.switchSignup
                        )}
                        onClick={toggleLoginHandler}
                        style={isLogin ? {} : hideForm}>
                        Don't have an account yet? <b>Signup</b>
                    </a>
                    <a
                        href="#/"
                        className={concatClasses(
                            styles.underlineHover,
                            styles.switchSignin
                        )}
                        onClick={toggleLoginHandler}
                        style={isLogin ? hideForm : {}}>
                        Already registered? <b>Loign</b>
                    </a>
                </div>
            </div>
        </div>
    );
};

export default Login;
