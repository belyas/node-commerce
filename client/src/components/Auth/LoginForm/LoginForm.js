import React from 'react';
import PropTypes from 'prop-types';

import { concatClasses } from '../../../utils/helpers';
import styles from '../Auth.module.css';

const LoginForm = ({
    isLogin,
    errors,
    email,
    password,
    error,
    blurHandler,
    changeHandler,
    loginSubmitHanlder,
    hideForm,
}) => {
    return (
        <form onSubmit={loginSubmitHanlder} style={isLogin ? {} : hideForm}>
            <h2
                className={concatClasses(
                    styles.fadeIn,
                    styles.first,
                    styles.formTitle
                )}>
                Login
            </h2>
            <input
                type="text"
                className={concatClasses(styles.fadeIn, styles.second)}
                name="email"
                placeholder="Email address"
                value={email}
                onChange={changeHandler}
                onBlur={blurHandler}
                required
            />
            {errors.email && (
                <p className="alert email-error alert-danger">{errors.email}</p>
            )}
            <input
                type="password"
                className={concatClasses(styles.fadeIn, styles.third)}
                name="password"
                placeholder="Password"
                value={password}
                onChange={changeHandler}
                onBlur={blurHandler}
                required
            />
            {errors.password && (
                <p className="alert password-error alert-danger">
                    {errors.password}
                </p>
            )}
            <input
                type="submit"
                className={concatClasses(styles.fadeIn, styles.fourth)}
                value="Log In"
            />
            {error && <p className="alert alert-danger">{error}</p>}
        </form>
    );
};

LoginForm.propTypes = {
    isLogin: PropTypes.bool.isRequired,
    errors: PropTypes.object.isRequired,
    email: PropTypes.string,
    password: PropTypes.string,
    error: PropTypes.string,
    blurHandler: PropTypes.func.isRequired,
    changeHandler: PropTypes.func.isRequired,
    loginSubmitHanlder: PropTypes.func.isRequired,
};

export default LoginForm;
