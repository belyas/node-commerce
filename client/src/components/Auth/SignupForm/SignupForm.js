import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from 'reactstrap';

import { concatClasses } from '../../../utils/helpers';
import styles from '../Auth.module.css';

const SignupForm = ({
    isLogin,
    hideForm,
    errors,
    email,
    password,
    firstname,
    lastname,
    error,
    changeHandler,
    blurHandler,
    signupSubmitHanlder,
    loading,
}) => {
    const submbitValue = loading ? <Spinner size="sm" /> : 'Register';

    return (
        <form onSubmit={signupSubmitHanlder} style={isLogin ? hideForm : {}}>
            <h2
                className={concatClasses(
                    styles.fadeIn,
                    styles.first,
                    styles.formTitle
                )}>
                Signup
            </h2>
            <input
                type="text"
                className={concatClasses(styles.fadeIn, styles.second)}
                name="firstname"
                placeholder="First name"
                value={firstname}
                onChange={changeHandler}
                onBlur={blurHandler}
                required
            />
            {errors.firstname && (
                <p className="alert firstname-error alert-danger">
                    {errors.firstname}
                </p>
            )}
            <input
                type="text"
                className={concatClasses(styles.fadeIn, styles.second)}
                name="lastname"
                value={lastname}
                placeholder="Last name"
                onChange={changeHandler}
                onBlur={blurHandler}
                required
            />
            {errors.lastname && (
                <p className="alert lastname-error alert-danger">
                    {errors.lastname}
                </p>
            )}
            <input
                type="text"
                className={concatClasses(styles.fadeIn, styles.third)}
                name="email"
                placeholder="Email address"
                onChange={changeHandler}
                onBlur={blurHandler}
                value={email}
                required
            />
            {errors.email && (
                <p className="alert email-error alert-danger">{errors.email}</p>
            )}
            <input
                type="password"
                className={concatClasses(styles.fadeIn, styles.third)}
                name="password"
                placeholder="Password"
                onChange={changeHandler}
                onBlur={blurHandler}
                value={password}
            />
            {errors.password && (
                <p className="alert password-error alert-danger">
                    {errors.password}
                </p>
            )}
            <button
                type="submit"
                className={concatClasses(styles.fadeIn, styles.fourth)}
                disabled={loading}>
                {submbitValue}
            </button>
            {error && <p className="alert alert-danger">{error}</p>}
        </form>
    );
};

SignupForm.propTypes = {
    isLogin: PropTypes.bool.isRequired,
    hideForm: PropTypes.object.isRequired,
    firstname: PropTypes.string,
    lastname: PropTypes.string,
    email: PropTypes.string,
    password: PropTypes.string,
    errors: PropTypes.object.isRequired,
    error: PropTypes.string,
    blurHandler: PropTypes.func.isRequired,
    changeHandler: PropTypes.func.isRequired,
    signupSubmitHanlder: PropTypes.func.isRequired,
    loading: PropTypes.bool,
};

export default SignupForm;
