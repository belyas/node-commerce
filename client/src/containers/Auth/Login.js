import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { validateLoginAuth } from '../../utils/validateAuth';
import { auth } from '../../store/actions';
import LoginComponent from '../../components/Auth/Login/Login';

class Login extends Component {
    state = {
        isLogin: true,
        errors: null,
        email: '',
        password: '',
    };

    toggleLoginHandler = () => {
        this.setState(prevState => {
            return {
                isLogin: !prevState.isLogin,
            };
        });
    };

    onSubmitHanlder = e => {
        e.preventDefault();
        this.props.onAuth(this.state.email, this.state.password);
    };

    onChangeHandler = event => {
        this.setState({
            [event.target.name]: event.target.value,
        });
    };

    onBlurHandler = () => {
        const validationErrors = validateLoginAuth({
            email: this.state.email,
            password: this.state.password,
        });

        this.setState({
            errors: validationErrors,
        });
    };

    render() {
        const props = {
            ...this.props,
            ...this.state,
        };
        return (
            <LoginComponent
                {...props}
                toggleLoginHandler={this.toggleLoginHandler}
                submitHanlder={this.onSubmitHanlder}
                changeHandler={this.onChangeHandler}
                blurHandler={this.onBlurHandler}
            />
        );
    }
}

Login.propTypes = {
    loading: PropTypes.bool.isRequired,
    isAuthenticated: PropTypes.bool.isRequired,
    error: PropTypes.string,
    token: PropTypes.string,
    onAuth: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
    return {
        loading: state.auth.loading,
        error: state.auth.error,
        isAuthenticated: state.auth.token !== null,
        token: state.auth.token,
    };
};

const mapDispatchToProps = dispatch => {
    return {
        onAuth: (email, password) => dispatch(auth(email, password)),
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Login);
