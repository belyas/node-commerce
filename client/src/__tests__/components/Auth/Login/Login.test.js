import React from 'react';
import { shallow, mount, render } from 'enzyme';

import Login from '../../../../components/Auth/Login/Login';
import LoginForm from '../../../../components/Auth/LoginForm/LoginForm';
import SignupForm from '../../../../components/Auth/SignupForm/SignupForm';

describe('Login component', () => {
    it('should verify the right rendered components and dom elements', () => {
        const props = {
            isAuthenticated: false,
            isLogin: true, // indicate the login form
            errors: {},
            email: null,
            password: null,
            error: null,
            blurHandler: () => ({}),
            changeHandler: () => ({}),
            loginSubmitHanlder: () => ({}),
            toggleLoginHandler: () => ({}),
            signupSubmitHanlder: () => ({}),
        };
        const wrapper = shallow(<Login {...props} />);
        const loginForm = wrapper.find('LoginForm');
        const signupForm = wrapper.find('SignupForm');

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.props().isAuthenticated).toBeFalsy();
        expect(loginForm).toHaveLength(1);
        expect(signupForm).toHaveLength(1);
        // signup form must be hidden
        expect(signupForm.prop('hideForm')).toEqual({ display: 'none' });
        expect(wrapper.find('a')).toHaveLength(2);
        expect(
            wrapper
                .find('a')
                .first()
                .text()
        ).toEqual("Don't have an account yet? Signup");
        expect(
            wrapper
                .find('a')
                .last()
                .text()
        ).toEqual('Already registered? Loign');
    });

    it('should redirect if isAuthenticated is set to true', () => {
        const props = {
            isAuthenticated: true,
            isLogin: true, // indicate the login form
            errors: {},
            email: null,
            password: null,
            error: null,
            blurHandler: () => ({}),
            changeHandler: () => ({}),
            loginSubmitHanlder: () => ({}),
            toggleLoginHandler: () => ({}),
            signupSubmitHanlder: () => ({}),
        };
        const wrapper = shallow(<Login {...props} />);

        // this is basically complaining about using Redirect in Login component
        expect(() => wrapper.html()).toThrow();
    });

    it('should display Login form component', () => {
        const props = {
            isLogin: true, // indicate the login form
            errors: {},
            email: 'yassine.belkaid87@gmail.com',
            password: '123456',
            error: null,
            blurHandler: () => ({}),
            changeHandler: () => {
                wrapper.setProps({ email: 'new.email@example.com' });
            },
            loginSubmitHanlder: () => {
                wrapper.setProps({ error: 'Invalid email or password' });
            },
        };
        const wrapper = shallow(<LoginForm {...props} />);

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.find('h2').text()).toEqual('Login');
        expect(wrapper.find('input')).toHaveLength(3);

        wrapper.simulate('submit');
        expect(wrapper.find('.alert-danger').text()).toEqual(
            'Invalid email or password'
        );

        wrapper.find('input[name="email"]').simulate('change');
        const currentHTML = wrapper.html();
        expect(currentHTML).not.toContain(props.email);
        expect(currentHTML).toContain('new.email@example.com');
    });

    it('should display signup form component', () => {
        const props = {
            isLogin: false, // indicate the signup form is displayed
            errors: {},
            email: 'yassine.belkaid87@gmail.com',
            password: '',
            firstname: 'yassine',
            lastname: 'belkaid',
            error: null,
            hideForm: {},
            blurHandler: () => {
                wrapper.setProps({ errors: { password: 'Password required' } });
            },
            changeHandler: () => {
                wrapper.setProps({ firstname: 'alex', lastname: 'wolf' });
            },
            signupSubmitHanlder: () => {
                wrapper.setProps({ error: 'Something went wrong!' });
            },
        };
        const wrapper = shallow(<SignupForm {...props} />);

        expect(wrapper.html()).toMatchSnapshot();

        expect(wrapper.find('h2').text()).toEqual('Signup');
        expect(wrapper.find('input')).toHaveLength(4);
        expect(wrapper.find('button')).toHaveLength(1);

        wrapper.find('input[name="firstname"]').simulate('change');
        wrapper.find('input[name="lastname"]').simulate('change');
        const currentHTML = wrapper.html();
        expect(currentHTML).toContain('alex');
        expect(currentHTML).toContain('wolf');

        wrapper.simulate('submit');
        expect(wrapper.find('.alert-danger').text()).toBe(
            'Something went wrong!'
        );

        wrapper.find('input[name="password"]').simulate('blur');
        expect(wrapper.find('.password-error').text()).toBe(
            'Password required'
        );
    });
});
