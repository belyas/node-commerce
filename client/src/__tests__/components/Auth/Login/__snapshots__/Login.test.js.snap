// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should display Login form component 1`] = `"<form><h2 class=\\"fadeIn first formTitle\\">Login</h2><input type=\\"text\\" class=\\"fadeIn second\\" name=\\"email\\" placeholder=\\"Email address\\" value=\\"yassine.belkaid87@gmail.com\\" required=\\"\\"/><input type=\\"password\\" class=\\"fadeIn third\\" name=\\"password\\" placeholder=\\"Password\\" value=\\"123456\\" required=\\"\\"/><input type=\\"submit\\" class=\\"fadeIn fourth\\" value=\\"Log In\\"/></form>"`;

exports[`Login component should display signup form component 1`] = `"<form><h2 class=\\"fadeIn first formTitle\\">Signup</h2><input type=\\"text\\" class=\\"fadeIn second\\" name=\\"firstname\\" placeholder=\\"First name\\" value=\\"yassine\\" required=\\"\\"/><input type=\\"text\\" class=\\"fadeIn second\\" name=\\"lastname\\" value=\\"belkaid\\" placeholder=\\"Last name\\" required=\\"\\"/><input type=\\"text\\" class=\\"fadeIn third\\" name=\\"email\\" placeholder=\\"Email address\\" value=\\"yassine.belkaid87@gmail.com\\" required=\\"\\"/><input type=\\"password\\" class=\\"fadeIn third\\" name=\\"password\\" placeholder=\\"Password\\" value=\\"\\"/><button type=\\"submit\\" class=\\"fadeIn fourth\\">Register</button></form>"`;

exports[`Login component should verify the right rendered components and dom elements 1`] = `"<div class=\\"wrapper fadeInDown\\"><div class=\\"autoMg formContent\\"><form><h2 class=\\"fadeIn first formTitle\\">Login</h2><input type=\\"text\\" class=\\"fadeIn second\\" name=\\"email\\" placeholder=\\"Email address\\" required=\\"\\"/><input type=\\"password\\" class=\\"fadeIn third\\" name=\\"password\\" placeholder=\\"Password\\" required=\\"\\"/><input type=\\"submit\\" class=\\"fadeIn fourth\\" value=\\"Log In\\"/></form><form style=\\"display:none\\"><h2 class=\\"fadeIn first formTitle\\">Signup</h2><input type=\\"text\\" class=\\"fadeIn second\\" name=\\"firstname\\" placeholder=\\"First name\\" required=\\"\\"/><input type=\\"text\\" class=\\"fadeIn second\\" name=\\"lastname\\" placeholder=\\"Last name\\" required=\\"\\"/><input type=\\"text\\" class=\\"fadeIn third\\" name=\\"email\\" placeholder=\\"Email address\\" required=\\"\\"/><input type=\\"password\\" class=\\"fadeIn third\\" name=\\"password\\" placeholder=\\"Password\\"/><button type=\\"submit\\" class=\\"fadeIn fourth\\">Register</button></form><div class=\\"formFooter\\"><a href=\\"#/\\" class=\\"underlineHover switchSignup\\">Don&#x27;t have an account yet? <b>Signup</b></a><a href=\\"#/\\" class=\\"underlineHover switchSignin\\" style=\\"display:none\\">Already registered? <b>Loign</b></a></div></div></div>"`;
